File: problem_66/diophantine_equation.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <numeric>

inline bool Is_Square(long long num){
	
	long long sqrt = std::sqrt(num);
	
	return (sqrt * sqrt == num);
}

std::pair<long long, long long> Get_Convergant_Fraction(std::vector<long long>& values){
	
	if(values.size() == 1) return {values[0], 1};
	
	long long denominator = values.back();
	long long numerator = 1;

	for(int i = static_cast<int>(values.size()) - 2; i > 0; i--){
		
		numerator += values[i] * denominator;
		long long gcd = std::gcd(numerator, denominator);
		numerator /= gcd;
		denominator /= gcd;

		std::swap(numerator, denominator);
	}

	numerator += denominator * values[0];

	long long gcd = std::gcd(numerator, denominator);
	numerator /= gcd;
	denominator /= gcd;

	std::cout<<numerator<<" "<<denominator<<std::endl;

	return {numerator, denominator};
}

long long Diophantine_Equation(){
	
	const int upper_limit = 2;

	for(long long d = 2; d <= upper_limit; d++){
				
		long long a = static_cast<long long>(std::sqrt(d));
		long long numerator = -a;
		long long factor = 1;
		long long denomirator = 1;
		
		if(a * a == d || d == 421) continue;
		
		std::vector<long long> a_values = {a};
		if(a * a - d == 1){
			std::cout<<d<<" : "<<a<<std::endl;
		} 
		std::cout<<a<<std::endl;
		while(true){
			
			long long rationalized_denomirator = factor * factor * d + numerator * numerator * -1;
			factor *= denomirator; 
			numerator = denomirator * -numerator;
			denomirator = rationalized_denomirator;
			
			long long gcd = std::gcd(factor, std::gcd(numerator, denomirator));

			factor /= gcd;
			numerator /= gcd;
			denomirator /= gcd;

			a = static_cast<long long>(static_cast<double>(std::sqrt(d) * factor + numerator)/static_cast<double>(denomirator));
			a_values.push_back(a);
			std::cout<<a<<std::endl;
			numerator -= a * denomirator;
			
			std::pair<long long, long long> fraction = Get_Convergant_Fraction(a_values);
			
			if(fraction.first * fraction.first - d * fraction.second * fraction.second == 1){
				std::cout<<d<<" : "<<fraction.first<<std::endl;
				break;
			}
		}
	}
			
		
	return 0;
}


int main(void){
	
	std::cout<<Diophantine_Equation()<<std::endl;

	return 0;
}

File: problem_66/diophantine_equation.py
import math

def Is_Square(num):
    
    sqrt = int(math.sqrt(num))
    
    if sqrt * sqrt == num:
        return True
    else:
        return False

def Get_Convergant_Fraction(a_values, d):
    
    idx = len(a_values) - 1
    numerator = 1
    denominator = a_values[idx]
    idx -= 1
    while(idx > 0):

        numerator += a_values[idx] * denominator
        gcd = math.gcd(numerator, denominator)
        numerator //= gcd
        denominator //= gcd

        numerator, denominator = denominator, numerator
        idx -= 1
    
    numerator += denominator * a_values[0]
    gcd = math.gcd(numerator, denominator)
    numerator //= gcd
    denominator //= gcd

    if(numerator * numerator - d * denominator * denominator == 1):
        return numerator

    return -1


def Diophantine_Equation():
    
    upper_limit = 1000
    result = 0
    highest_seen = 0

    for d in range(2, upper_limit+1):
        
        if(Is_Square(d)):
            continue

        a = int(math.sqrt(d))
        numerator = -a
        denomirator = 1
        factor = 1
        a_values = [a]
        while(True):
            
            rationalized_denomirator = factor * factor * d - numerator * numerator
            factor *= denomirator
            numerator = denomirator * -numerator
            denomirator = rationalized_denomirator

            gcd = math.gcd(factor, math.gcd(numerator, denomirator))
            
            numerator //= gcd
            denomirator //= gcd
            factor //= gcd

            a = (factor * math.sqrt(d) + numerator) / denomirator
            a = int(a)
            
            numerator -= a * denomirator;

            a_values.append(a)
            fraction = Get_Convergant_Fraction(a_values, d)
            if fraction != -1:
                if fraction > highest_seen:
                    highest_seen = fraction
                    result = d
                break

    return result

print(Diophantine_Equation())
#661

